src/models/melody/melody.model.ts

```diff
@@ -0,0 +1,134 @@
+import * as Tone from 'tone';
+
+export type OnSeek = (index: number, playing: boolean) => void;
+
+export class Melody {
+  private synth = new Tone.Synth().toDestination();
+  private forwardSequence: Tone.Sequence;
+  private backwardSequence: Tone.Sequence;
+  private currentDatumIndex = 0;
+  private inclusive = true; // if true, playing the melody starting inclusively from currentDatumIndex
+  private reversed = false; // if true, playing the melody backward
+
+  constructor(
+    private values: number[],
+    private frequencyRange: [number, number],
+    private duration: number, // duration (in seconds) of the melody
+    private onSeek?: OnSeek,
+  ) {
+    const reversedValues = [...values].reverse();
+    this.forwardSequence = this.createSequence(values);
+    this.backwardSequence = this.createSequence(reversedValues);
+  }
+
+  get noteDuration() {
+    return this.duration / this.values.length;
+  }
+
+  get currentSeconds() {
+    return this.reversed
+      ? this.duration - Tone.Transport.seconds
+      : Tone.Transport.seconds;
+  }
+
+  get isPlaying() {
+    return Tone.Transport.state === 'started';
+  }
+
+  get isEnded() {
+    return (
+      this.reversed && this.currentDatumIndex === 0 ||
+      !this.reversed && this.currentDatumIndex === this.values.length - 1
+    );
+  }
+
+  get nextIndex() {
+    if (this.isEnded) {
+      return (this.values.length - 1) - this.currentDatumIndex; // bring playhead to the opposite end
+    }
+    const offset = this.inclusive ? 0 : (this.reversed ? -1 : +1);
+    return this.currentDatumIndex + offset;
+  }
+
+  private static getKeyNumber(frequency: number) {
+    return Math.log2(frequency / 440) * 12 + 49;
+  }
+
+  private static getFrequency(keyNumber: number) {
+    return Math.pow(2, (keyNumber - 49) / 12) * 440;
+  }
+
+  async resume(reversed: boolean) {
+    if (Tone.getContext().state === 'suspended') {
+      await Tone.start();
+    }
+    if (!this.isPlaying) {
+      this.reversed = reversed;
+      this.currentDatumIndex = this.nextIndex;
+      let nextSeconds = this.getSeconds(this.currentDatumIndex);
+      if (this.reversed) {
+        this.backwardSequence?.start(0);
+        this.forwardSequence?.stop(0);
+        nextSeconds += this.noteDuration / 2;
+        Tone.Transport.start(undefined, this.duration - nextSeconds);
+      } else {
+        this.backwardSequence?.stop(0);
+        this.forwardSequence?.start(0);
+        nextSeconds -= this.noteDuration / 2;
+        Tone.Transport.start(undefined, nextSeconds);
+      }
+    }
+  }
+
+  pause() {
+    if (this.isPlaying) {
+      Tone.Transport.pause();
+      this.seekTo(this.currentSeconds);
+    }
+  }
+
+  getCurrentDatumIndex() {
+    return this.currentDatumIndex;
+  }
+
+  seekTo(seconds: number, inclusive = false) {
+    this.currentDatumIndex = this.getDatumIndex(seconds);
+    this.inclusive = this.isEnded || inclusive;
+    this.onSeek?.(this.currentDatumIndex, this.isPlaying);
```

Comment:
Is this callback necessary? Could the LineChartAudification component instead just get the currentDatumIndex and isPlaying from this.melody and execute its logic in order to avoid the need to pass the callback down?

Reply:
I needed to pass the callback in order for the component to know **when** to update the active datum on the visualization and to read it out.